npeps/npep-122.md

```diff
@@ -158,7 +158,147 @@ Fourth, the ANP subject allows using pod selectors, while tenancy use cases only
 
 ## API
 
-TBD
+### Tenant definition
+
+We define Tenant as a set of namespaces.
+All Namespaces affected by Tenancy will be split into Tenants based on their label(s), we call these labels `tenancyLabels`. 
+A Tenant is identified by values of `tenancyLabels`, which are shared by all namespaces in the Tenant.
+
+There are 2 ways to select which namespaces should be affected by Tenancy:
+
+1. Separate from `tenancyLabels`. We have a separate label selector to define which namespaces should be affected by Tenancy.
+Then selected namespaces are split into Tenants.
+
+    **Cons**: selected namespace may not have some of the `tenancyLabels`, which will likely result in introducing "None" value
+    for `tenancyLabels`. To only select namespaces with existing labels extra `key: X, operator: Exists` condition should
+    be added to the selector.
+
+2. Joined with `tenancyLabels`. We select all namespace that all `tenancyLabels` present.
+
+    **Cons**: applying all labels from `tenancyLabels` to a namespaces automatically makes it affected by tenancy.
+    No way to make sure ALL namespaces in a cluster are a part of some tenant, e.g. newly created namespace will not
+    be limited/protected by Tenancy rules until it gets `tenancyLabels` assigned, may potentially be a security problem.
+    
+    Example: by setting deny-from-other-tenants I want to ensure only namespaces with the same label `security-zone` value
+    can send incoming connections. That policy will be violated if a new tenant is created without `security-zone` label,
+    meaning it is not a part of any Tenant. This may potentially be solved by using allow-from-same-tenant + deny ALL
+    ANP, but it is stricter, since it affects all incoming connections, not just from other pods.
+
+### Peers and actions
+
+Based on the existing User Stories, Tenancy only needs Allow and Deny actions, and the only 2 types of peers are
```

Comment:
I would disagree. I think user story 4.2 indicates that there is a need for PASS in case of strict tenant isolation, meaning "as a cluster admin I want to strictly disallow cross-tenant traffic, but owners of individual tenant namespaces can decide what they want to allow/deny in the scope of these namespaces"

Reply:
I am not sure I understand it correctly. So "strictly disallow cross-tenant traffic" means `Deny from/to notSameTenant` right? How would you combine it with PASS? maybe a yaml example would help me figure it out

Reply:
What I mean is, instead of "deny from/to not same tenant", admin might want to define policies in the lines of "pass from/to same tenant, and deny everything else", especially when the egress story comes into play

Reply:
I think we need a new use case for that, the semantics is "as a cluster admin, I want to let namespaces decide what is allowed within their tenant, but strictly deny everything else" and it doesn't match any of the existing use stories
it also can be expressed as "deny from not same tenant" + "ANP PASS empty pod selector", but it is less obvious for a given use case
 @danwinship wdyt?

Reply:
Tenancy-based rules are inherently east-west; you're always saying "traffic is allowed/denied between namespaces where these labels have the same/different values". You don't need to use a tenancy-based rule for N/S traffic, because the outside world can never be SameTenant.

IOW, if you want to implement "traffic is allowed within tenants and denied to/from everywhere else (including N/S)", then that would be:
1. a tenancy-based rule saying "traffic between tenants is denied"
2. a non-tenancy-based rule saying "N/S traffic to/from user namespaces is denied"

in which case there's still no need for a tenancy-based pass action.

----

Taking a step back, there are two concepts that we have sort of been mixing together:
1. "tenancy" - The idea that namespaces are grouped into tenants, and that you can use "same tenant" and "not same tenant" as selectors in policies.
2. "cluster default policy" - The idea that the administrator has some high-level plan for how networking works "normally" in the cluster (ie, when there aren't more-specific policies overriding it). eg:
    - "Traffic is permitted by default."
    - "Traffic within user namespaces is permitted by default, but traffic between user namespaces is blocked." 
    - "Traffic within the cluster is allowed by default, but cluster egress traffic is blocked." 

The concept of "cluster default policy" does not explicitly exist in ANP, but it's sort of lurking in the background, and all of the user stories for tenancy involve defining part of "cluster default policy", but tenancy isn't _all_ of "cluster default policy".

Part of the "tenancy is confusing / too powerful" problem that we're fixing now was because there was nothing stopping you from using tenancy-based rules in arbitrary ANPs/BANPs, even though the user stories only required tenancy-based rules in "cluster default policy" ANPs/BANPs... The new proposals below are better, since you can only define tenancy rules that apply to everyone, rather than randomly using a sameLabels rule inside a policy that only matches some traffic.

Reply:
> IOW, if you want to implement "traffic is allowed within tenants and denied to/from everywhere else (including N/S)"

Well my point is, I was thinking the usecase as "I don't care about the traffic within tenants, but I want to make sure traffic is denied to/from everywhere else (including N/S)". This also is not a "cluster default policy" IMO because inside a tenant I do not want to define any rule, rather than "Traffic within user namespaces is permitted by default" which sounded like a BANP.